### ISSUE TITLE
DERP bug: Transfer issues when one node is behind a "NAT"

### ISSUE THREAD

**Issue by ramfox** [n0-computer/iroh#1404]
Need a place to dump thoughts about this.

Now that we have a way to force a "derp only" mode, we can simulate what it is like when one or both sides of the data transfer is behind a nat.

A few problems have become apparent, specifically with the mix of transferring data over the relay while also attempting to hole punch using the derper & the UDP sockets to send pings.

Here are a few scenarios I have tested and the results:

### **1)  Provider and Getter are both in derp only mode, and neither attempts to do any hole-punching**
This scenario works! I was able to transfer 340mbs of an audiobook from one computer to the other. 

### **2) Provider and Getter are both in derp only mode, and only the Getter attempts to do hole-punching**
This error also happens if both the provider and getter are in derp only mode & they both attempt to do hole-punching, but to simplify the following explanation, it's easier to just think of the provider as behind a nat and the getter as the one attempting hole punching.
In this scenario, the Getter has all of the Provider's addresses (from the ticket), and is able to send pings to the Provider.
I haven't investigated this fully, but the last error messages are consistently:
```
2023-08-25T01:30:36.455440Z TRACE drive{id=0}: quinn_proto::connection: timeout timer=KeepAlive
2023-08-25T01:30:36.455594Z TRACE drive{id=0}: quinn_proto::connection: sending keep-alive
2023-08-25T01:30:36.456119Z TRACE drive{id=0}:send{space=Data pn=2832}: quinn_proto::connection: PING
2023-08-25T01:30:36.456261Z TRACE drive{id=0}:send{space=Data pn=2832}: quinn_proto::connection::packet_builder: PADDING * 2
Receiving 'The Empty Grave Lockwood & Co., Book 5 (Unabridged).mp3'...
⠈ [00:00:23] [#>-------------------------------------------------------------------------------------------------------------------------------------------------------------] 3.69 MiB/348.62 MiB (2262.3s)2023-08-25T01:30:37.458059Z TRACE drive{id=0}: quinn_proto::connection: timeout timer=KeepAlive
2023-08-25T01:30:37.458109Z TRACE drive{id=0}: quinn_proto::connection: sending keep-alive
2023-08-25T01:30:37.458275Z TRACE drive{id=0}:send{space=Data pn=2833}: quinn_proto::connection: PING
2023-08-25T01:30:37.458322Z TRACE drive{id=0}:send{space=Data pn=2833}: quinn_proto::connection::packet_builder: PADDING * 2
Receiving 'The Empty Grave Lockwood & Co., Book 5 (Unabridged).mp3'...
⠦ [00:00:25] [#>-------------------------------------------------------------------------------------------------------------------------------------------------------------] 3.69 MiB/348.62 MiB (2450.0s)2023-08-25T01:30:38.460542Z TRACE drive{id=0}: quinn_proto::connection: timeout timer=KeepAlive
2023-08-25T01:30:38.460726Z TRACE drive{id=0}: quinn_proto::connection: sending keep-alive
2023-08-25T01:30:38.461105Z TRACE drive{id=0}:send{space=Data pn=2834}: quinn_proto::connection: PING
2023-08-25T01:30:38.461201Z TRACE drive{id=0}:send{space=Data pn=2834}: quinn_proto::connection::packet_builder: PADDING * 2
Receiving 'The Empty Grave Lockwood & Co., Book 5 (Unabridged).mp3'...
⠒ [00:00:26] [#>-------------------------------------------------------------------------------------------------------------------------------------------------------------] 3.69 MiB/348.62 MiB (2699.0s)2023-08-25T01:30:39.462176Z TRACE drive{id=0}: quinn_proto::connection: timeout timer=KeepAlive
2023-08-25T01:30:39.462259Z TRACE drive{id=0}: quinn_proto::connection: sending keep-alive
2023-08-25T01:30:39.462480Z TRACE drive{id=0}:send{space=Data pn=2835}: quinn_proto::connection: PING
2023-08-25T01:30:39.462555Z TRACE drive{id=0}:send{space=Data pn=2835}: quinn_proto::connection::packet_builder: PADDING * 2
Receiving 'The Empty Grave Lockwood & Co., Book 5 (Unabridged).mp3'...
⠈ [00:00:27] [#>-------------------------------------------------------------------------------------------------------------------------------------------------------------] 3.69 MiB/348.62 MiB (3001.9s)2023-08-25T01:30:40.464313Z TRACE drive{id=0}: quinn_proto::connection: timeout timer=KeepAlive
2023-08-25T01:30:40.464419Z TRACE drive{id=0}: quinn_proto::connection: sending keep-alive
2023-08-25T01:30:40.464793Z TRACE drive{id=0}:send{space=Data pn=2836}: quinn_proto::connection: PING
2023-08-25T01:30:40.464910Z TRACE drive{id=0}:send{space=Data pn=2836}: quinn_proto::connection::packet_builder: PADDING * 2
Receiving 'The Empty Grave Lockwood & Co., Book 5 (Unabridged).mp3'...
⠒ [00:00:27] [#>-------------------------------------------------------------------------------------------------------------------------------------------------------------] 3.69 MiB/348.62 MiB (3132.0s)^C2023-08-25T01:30:40.837023Z TRACE quinn_proto::connection: connection closed
2023-08-25T01:30:40.837398Z TRACE drive{id=0}:send{space=Data pn=2837}: quinn_proto::connection: sending CONNECTION_CLOSE
2023-08-25T01:30:40.837562Z TRACE drive{id=0}:send{space=Data pn=2837}: quinn_proto::connection: ACK ArrayRangeSet([2448..3437]), Delay = 0us
2023-08-25T01:30:40.837851Z TRACE drive{id=0}: quinn_proto::connection: sending 37 bytes in 1 datagrams
```
The Getter closes the connection & terminates the program.
The Provider side gets a message that the connection as closed, it prints the error, but it stays running.

These `quinn_proto::connection: timeout timer=KeepAlive` warnings are never triggered in scenario 1

My only current theory is:
- There is some blocking that occurs on the derp actor during hole punching attempts, and it causes the data to stop flowing as expected, causing the quinn endpoint to believe there is connection loss occurring.
- For whatever reason, this blockage does not occur on the main magicsock actor, and so when we have a direct connection over the UDP socket, this data is not blocked, and so the quinn endpoint does not view the hole punching attempts as connection loss.

### **3) Provider is in derp only mode (which does no hole punching), and the Getter is in normal mode
Same as scenario 2

**Comment by flub** [comment]
I don't understand what happens with holepunching when one of the nodes is in derp-only mode.  suppose you punch a hole, what then?  doesn't get used?

**Comment by dignifiedquire** [comment]
@ramfox you missed pasting in the error message 😅

**Comment by ramfox** [comment]
> I don't understand what happens with holepunching when one of the nodes is in derp-only mode. suppose you punch a hole, what then? doesn't get used?

Yes, from the Provider's side, we will always just choose the DERP relay whenever we have a message to send & we never attempt to hole punch or respond to hole punching attempts.

The variation comes from Getter side:

In scenario 1, we don't do any hole punching on either node, so no pings or call me maybe messages get initiated ever.

In scenario 2, the Provider does not respond to or initiate any hole-punching pings or call me maybe messages. But the Getter sends out pings and call me maybes. However, the Getter will never use any of those addresses to send packets it gets from the quinn endpoint, it will always choose to send over the DERP relay. 

In scenario 3, the Provider does not respond to or initiate any hole-punching pings or call me maybe message. The Getter runs in normal mode, so it attempts to hole punch and attempts to send messages to the Provider via UDP and DERP.

**Comment by ramfox** [comment]
> @ramfox you missed pasting in the error message 😅

fixed, lmao

**Comment by ramfox** [comment]
Update:
- looks like there is an error on the derper which closes conenctions to the client. Sometimes we reconnect, sometimes we don't. Why do the reconnects that fail fail?
- need to see if these errors occur when in "derp only" mode & we just don't notice them because we connect fast enough or if the errors occur *because* or *during* hole punching attempts

**Comment by ramfox** [comment]
further investigation showed that we were getting an "Unknown Frame Type", which was causing a panic in quinn.

This was fixed in quinn, and updated in iroh in #1510 

So we should no longer have the panic, but we still don't know what frame / issue is causing an "unknown frame type" to propagate through our network

**Comment by dignifiedquire** [comment]
@ramfox can we close this?
